client/state/happychat/middleware.js

```diff
@@ -67,6 +132,14 @@ export default function( connection = null ) {
 
 	return store => next => action => {
 		switch ( action.type ) {
+			case HAPPYCHAT_CONNECTING:
+				if ( getHappychatConnectionStatus( store.getState() ) === 'connected' ) {
+					// If chat is already connected, do nothing and stop the action from proceeding to reducers
+					return;
```

Comment:
Thanks for the detailed explanation. That definitely helps me understand the rational here :)
As far as I know, our current practice of dealing with duplicated requests like this is via query components. e.g. [QueryCountryStates](https://github.com/Automattic/wp-calypso/blob/master/client/components/data/query-country-states/index.jsx). You might want to look into its `request` method. The essence of this approach is actually the first approach from your list: prevent components from dispatching if already connected, or in the case of this query component, if the request is in progress.  You might ask: people can still access the underlying action creators and ruin the purpose. That's totally true, and the way we nib them in the buds is what you might expect already: code review :)

However, that doesn't mean I'm against your approach. As we moving more and more this kind of async data flow to the middleware, I'm having the feeling that your approach might be better in several ways:

1. When calling an action creator, developers no longer need to search if there is a corresponding query component.
1. Having a unified place to do this kind of delicate manipulation means we have a better chance to refactor out common code, whether buffering / de-dupe request / etc. Actually, I believe that is what [wpcom-http](https://github.com/Automattic/wp-calypso/tree/master/client/state/data-layer/wpcom-http) middleware ultimately aiming for.